pkg/providers/vsphere/controlplane_test.go

```diff
@@ -125,6 +125,46 @@ func TestControlPlaneSpecNoKubeVersion(t *testing.T) {
 	g.Expect(err).To(MatchError(ContainSubstring("generating vsphere control plane yaml spec")))
 }
 
+func TestControlPlaneSpecUpdateMachineTemplates(t *testing.T) {
+	g := NewWithT(t)
+	logger := test.NewNullLogger()
+	ctx := context.Background()
+	spec := givenClusterSpec(t, testClusterConfigMainFilename)
+	originalKubeadmControlPlane := kubeadmControlPlane()
+	originalEtcdCluster := etcdCluster()
+	originalEtcdCluster.Spec.InfrastructureTemplate.Name = "test-etcd-2"
+	client := test.NewFakeKubeClient(
+		originalKubeadmControlPlane,
+		originalEtcdCluster,
+		vsphereMachineTemplate("test-control-plane-1"),
+		vsphereMachineTemplate("test-etcd-2"),
+	)
+	spec.VSphereMachineConfigs["test-etcd"].Spec.Datastore = "new-datastore"
+
+	kcp := wantKCP()
+	kcp.Spec.MachineTemplate.InfrastructureRef.Name = "test-control-plane-2"
+
+	etcd := wantEtcdCluster()
+	etcd.Spec.InfrastructureTemplate.Name = "test-etcd-3"
+
+	wantCPtemplate := vsphereMachineTemplate("test-control-plane-2")
```

Comment:
could we do a deep copy of the original one instead of generating it again?
same for the etcd one

I think it helps understand your intent: you want the same object and modify the expected changes on top